Change handling of single/multi thread computation of PostProccess filtering + more tests

maxwest-uw opened the pull request:
- changed default number of cores from 30 to 1
- made it so that if the number of cores is set to 1, PostProcess will calculate filters single threaded for a speed boost
- added unit tests for filters in both single and multi thread

jeremykubica commented on tests/test_filtering.py:
```diff
@@ -1,9 +1,111 @@
+            time = i/self.imCount
+            im = layered_image(str(i), self.dim_x, self.dim_y, 
+                                self.noise_level, self.variance, time,
+                                self.p)
```
Small comment: There is an extra space at the start of lines 74 and 75. They should be aligned with str(i).

jeremykubica commented on tests/test_filtering.py:
```diff
@@ -1,9 +1,111 @@
+                                self.p)
+            im.add_object(self.start_x + time*self.x_vel+0.5,
+                           self.start_y + time*self.y_vel+0.5,
+                           self.object_flux)
```
Small comment: There is an extra space at the start of lines 77 and 78. They should be aligned with self.start_x.

jeremykubica commented on tests/test_filtering.py:
```diff
@@ -1,9 +1,111 @@
+        self.search = stack_search(self.stack)
+        self.search.search(self.angle_steps, self.velocity_steps,
+                         self.min_angle, self.max_angle, self.min_vel,
+                         self.max_vel, int(self.imCount/2))
```
Align 99 and 100 with line 98 (add two spaces).

jeremykubica commented on analysis/analysis_utils.py:
```diff
@@ -999,23 +1014,35 @@ def apply_stamp_filter(
             keep['final_results'] = np.unique(np.concatenate(passing_stamps_idx))
         print('Keeping %i results' % len(keep['final_results']), flush=True)
+        end_time = time.time()
+        time_elapsed = end_time-start_time
```
Add spaces around minus: "end_time-start_time" -> "end_time - start_time"

jeremykubica commented on tests/test_filtering.py:
```diff
@@ -1,9 +1,111 @@
+
+        # search parameters
+        self.angle_steps = 150
+        self.velocity_steps = 150
```
Given that you know the object's real trajectory, you might be able to reduce the number of steps here and speed up the test. 150 is a reasonably large number of steps.

jeremykubica commented on tests/test_filtering.py:
```diff
@@ -1,9 +1,111 @@
+        
+        self.imlist = []
+        for i in range(self.imCount):
+            time = i/self.imCount
```
Add a space around the division sign: "i/self.imCount" -> "i / self.imCount"

jeremykubica commented on tests/test_filtering.py:
```diff
@@ -1,9 +1,111 @@
+            im = layered_image(str(i), self.dim_x, self.dim_y, 
+                                self.noise_level, self.variance, time,
+                                self.p)
+            im.add_object(self.start_x + time*self.x_vel+0.5,
```
Add spaces around the mathematical operators.

jeremykubica commented on tests/test_filtering.py:
```diff
@@ -1,9 +1,111 @@
+        
+        # Set up old_results object for analysis_utils.PostProcess
+        psi_curves = [[1., 1., 1.], [1., 1., 1.], [1., 1., 1.]]
+        phi_curves = [[1., 1., 1.], [1., 1., 1.], [1., 1., 1.]]
```
Would it be worth adding a 4th example that gets rejected by the filters?

jeremykubica commented on tests/test_filtering.py:
```diff
@@ -200,7 +302,62 @@ def test_calculate_likelihood_psiphi_zero_or_negative_phi(self):
+        
+        self.assertEqual(len(res), 3)
+        for r in res:
+            self.assertEqual(r[1], [-1])
```
I'm not sure that I'm following this assert. I'm probably missing something. What is r[1]? From the code it looks like it is the indices that pass the filtering. But, if that's the case, how does it equal -1?

maxwest-uw replied:
[-1] is the list of passing indices for a given psi/phi curve when it's rejected, all of the results for these particular test curves are filtered out because that's easier to test but I think you're right that there should be a mix of passes and rejections for each test.

jeremykubica replied:
If we could include a curve where not all results are filtered (it could have all results pass), that would give us better coverage on the tests.

jeremykubica reviewed (commented):
This is a really nice change. I've added a bunch of stylistic comments.

The only blocking comment is the question in test_filtering.py on line 313. I want to make sure I understand what the tests are doing.

maxwest-uw commented:
note: I looked back over the tests and thought it was weird that the clipped filtering filters out psi and phi curves that are all the same values.  i figured out why this is happening with clipped_average and changed the test data to reflect this but I'm still working on tweaking it for the other ones. it had to do with how the filtering removes the 'clipped' elements

maxwest-uw commented:
changed the filtering tests to get better coverage, added more comments, and moved the search for stamp filtering into the actual stamp filtering tests for a speed up :)

jeremykubica reviewed (approved):
This looks good. Thanks for expanding the test coverage.
